velox/functions/lib/DateTimeImpl.h

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <velox/type/Timestamp.h>
+#include "velox/external/date/date.h"
+#include "velox/functions/lib/DateTimeFormatter.h"
+#include "velox/functions/lib/DateTimeImpl.h"
+#include "velox/functions/lib/TimeUtils.h"
+
+namespace facebook::velox::functions {
+
+// Returns toTimestamp - fromTimestamp expressed in terms of unit.
+FOLLY_ALWAYS_INLINE int64_t diffTimestamp(
+    const DateTimeUnit unit,
+    const Timestamp& fromTimestamp,
+    const Timestamp& toTimestamp) {
+  // TODO(gaoge): Handle overflow and underflow with 64-bit representation
+  if (fromTimestamp == toTimestamp) {
+    return 0;
+  }
+
+  const int8_t sign = fromTimestamp < toTimestamp ? 1 : -1;
+
+  // Spark support microsecond unit, if convert by .toMicros() for all the
+  // unit, may cause overflow, so treat it dependently.
```

Comment:
Sorry, I did not understand this comment. Would you please clarify a bit?

Reply:
Presto UT Timestamp(9223372036854775, 0) will overflow when do  `std::min(fromTimestamp, toTimestamp).toMicros())`.